File: apps/api/models/connections.py
"""Document-connections wire models (ADR-027 D1 — ROADMAP E4 exploration surface).

One document's **neighbourhood**: semantic neighbours plus the documents that cite it. The
outgoing side — the paper's own reference list — is served by ``references`` instead
(2026-08-10): a reader wants one bibliography in one place, and duplicating its resolved
half here made the two blocks disagree about what the paper cites.
"""

from __future__ import annotations

from typing import TYPE_CHECKING

from pydantic import BaseModel

if TYPE_CHECKING:
    from doc_assistant.library import DocConnections


class RelatedDocPayload(BaseModel):
    """One semantic neighbour (mirrors ``library.SimilarDoc``). ``score`` = cosine."""

    document_id: str
    filename: str
    title: str | None
    score: float


class CitedByPayload(BaseModel):
    """One in-corpus document citing the subject (deduped; mirrors ``library.CitedByDoc``)."""

    document_id: str
    filename: str
    n_citations: int


class DocConnectionsPayload(BaseModel):
    """A document's exploration bundle (E4): semantic neighbours + the documents citing it.

    What this document *cites* is the References block's payload (``references``), not
    this one."""

    related: list[RelatedDocPayload]
    cited_by: list[CitedByPayload]

    @classmethod
    def from_bundle(cls, b: DocConnections) -> DocConnectionsPayload:
        return cls(
            related=[
                RelatedDocPayload(
                    document_id=r.target_document_id,
                    filename=r.target_filename,
                    title=r.target_title,
                    score=r.score,
                )
                for r in b.related
            ],
            cited_by=[
                CitedByPayload(
                    document_id=d.document_id,
                    filename=d.filename,
                    n_citations=d.n_citations,
                )
                for d in b.cited_by
            ],
        )

File: apps/desktop/src/lib/core/types/connections.ts
// TypeScript mirror of the desktop-API payloads (apps/api/models/connections.py).
// Keep in sync with the pydantic models — this is the wire contract; a change to the
// model and a change here belong in the same commit (apps/desktop/CLAUDE.md).
//
// Document connections (ADR-027 D1, E4).
// Mirrors apps/api/models/connections.py.

// Document connections (ADR-027 D1 — E4 exploration surface). GET
// /api/library/documents/{id}/connections. Mirrors apps/api/models/connections.py::DocConnectionsPayload.
//
// This document's NEIGHBOURHOOD — not its bibliography. What the paper cites lives in
// `references.ts` / the References block; keeping the resolved half here too made the two
// blocks disagree about the same document.
export interface RelatedDoc {
  document_id: string
  filename: string
  title: string | null
  score: number
}
export interface CitedByDoc {
  document_id: string
  filename: string
  n_citations: number
}
export interface DocConnections {
  related: RelatedDoc[]
  cited_by: CitedByDoc[]
}
